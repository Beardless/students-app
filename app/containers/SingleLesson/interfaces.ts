export interface LessonInterface {
  id: string;
  title: string;
  description: string;
  position: number;
  embed_code: string;
  created: string;
  updated: string;
  module: {
    id: string;
    title: string;
    course: {
      id: number;
      title: string;
    };
  };
}
