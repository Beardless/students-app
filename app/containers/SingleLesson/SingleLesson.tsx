import * as React from "react";
import CourseLesson from "../../components/CourseLesson";
import { match } from "react-router";
import { LessonInterface } from "./interfaces";
import LoadingIndicator from "../../components/LoadingIndicator";

interface SingleLessonProps {
  getLesson: (lessonId: string) => void;
  error: boolean | string;
  loading: boolean;
  currentLesson: LessonInterface;
  match: match<{ lessonId: string }>;
}

class SingleLesson extends React.Component<SingleLessonProps> {
  componentDidMount() {
    const {
      params: { lessonId }
    } = this.props.match;
    this.props.getLesson(lessonId);
  }

  render(): React.ReactNode {
    const { error, loading, currentLesson } = this.props;

    if (error) {
      return <div>{error}</div>;
    }
    if (loading) {
      return <LoadingIndicator />;
    }

    return <CourseLesson item={currentLesson} />;
  }
}

export default SingleLesson;
