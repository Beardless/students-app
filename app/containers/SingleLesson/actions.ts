import {
  LESSON_ACTION,
  LESSON_ACTION_SUCCESS,
  LESSON_ACTION_ERROR,
  LESSON_COMPLETE_ACTION,
  LESSON_COMPLETE_ACTION_SUCCESS,
  LESSON_COMPLETE_ACTION_ERROR
} from "./constants";
import { LessonInterface } from "./interfaces";

export function lessonAction(
  id: string
): { type: string; payload: { id: string } } {
  return {
    type: LESSON_ACTION,
    payload: { id }
  };
}

export function lessonActionSuccess(
  lesson: LessonInterface
): { type: string; lesson: LessonInterface } {
  return {
    type: LESSON_ACTION_SUCCESS,
    lesson
  };
}

export function lessonActionError(
  error: string
): { type: string; error: string } {
  return {
    type: LESSON_ACTION_ERROR,
    error
  };
}

export function lessonCompleteAction(
  isComplete: boolean,
  lessonId: string
): { type: string; payload: { isComplete: boolean; lessonId: string } } {
  return {
    type: LESSON_COMPLETE_ACTION,
    payload: {
      isComplete,
      lessonId
    }
  };
}

export function lessonCompleteActionSuccess(
  lesson: LessonInterface
): { type: string; lesson: LessonInterface } {
  return {
    type: LESSON_COMPLETE_ACTION_SUCCESS,
    lesson
  };
}

export function lessonCompleteActionError(
  error: string
): { type: string; error: string } {
  return {
    type: LESSON_COMPLETE_ACTION_ERROR,
    error
  };
}
