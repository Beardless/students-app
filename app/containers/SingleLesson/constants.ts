//Lesson action
export const LESSON_ACTION: string = "courses/SingleLesson/LESSON_ACTION";
export const LESSON_ACTION_SUCCESS: string =
  "courses/SingleLesson/LESSON_ACTION_SUCCESS";
export const LESSON_ACTION_ERROR: string =
  "courses/SingleLesson/LESSON_ACTION_ERROR";

//Lesson complete action
export const LESSON_COMPLETE_ACTION: string =
  "courses/SingleLesson/LESSON_COMPLETE_ACTION";
export const LESSON_COMPLETE_ACTION_SUCCESS: string =
  "courses/SingleLesson/LESSON_COMPLETE_ACTION_SUCCESS";
export const LESSON_COMPLETE_ACTION_ERROR: string =
  "courses/SingleLesson/LESSON_COMPLETE_ACTION_ERROR";

//Create bookmark action
export const CREATE_BOOKMARK_ACTION: string =
  "courses/SingleLesson/CREATE_BOOKMARK_ACTION";
export const CREATE_BOOKMARK_ACTION_SUCCESS: string =
  "courses/SingleLesson/CREATE_BOOKMARK_ACTION_SUCCESS";
export const CREATE_BOOKMARK_ACTION_ERROR: string =
  "courses/SingleLesson/CREATE_BOOKMARK_ACTION_ERROR";

//Get bookmark list action
export const GET_BOOKMARK_LIST_ACTION: string =
  "courses/SingleLesson/GET_BOOKMARK_LIST_ACTION";
export const GET_BOOKMARK_LIST_ACTION_SUCCESS: string =
  "courses/SingleLesson/GET_BOOKMARK_LIST_ACTION_SUCCESS";
export const GET_BOOKMARK_LIST_ACTION_ERROR: string =
  "courses/SingleLesson/GET_BOOKMARK_LIST_ACTION_ERROR";

//Delete bookmark action
export const DELETE_BOOKMARK_ACTION: string =
  "courses/SingleLesson/DELETE_BOOKMARK_ACTION";
export const DELETE_BOOKMARK_ACTION_SUCCESS: string =
  "courses/SingleLesson/DELETE_BOOKMARK_ACTION_SUCCESS";
export const DELETE_BOOKMARK_ACTION_ERROR: string =
  "courses/SingleLesson/DELETE_BOOKMARK_ACTION_ERROR";
