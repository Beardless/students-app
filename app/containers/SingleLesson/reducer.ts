import { fromJS } from "immutable";

import {
  LESSON_ACTION,
  LESSON_ACTION_SUCCESS,
  LESSON_ACTION_ERROR,
  LESSON_COMPLETE_ACTION,
  LESSON_COMPLETE_ACTION_SUCCESS,
  LESSON_COMPLETE_ACTION_ERROR
} from "./constants";

const initialState = fromJS({
  loading: false,
  completeLoading: false,
  completeError: false,
  error: false,
  currentLesson: false
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LESSON_ACTION:
      return state.set("loading", true).set("error", false);
    case LESSON_ACTION_SUCCESS:
      return state.set("loading", false).set("currentLesson", action.lesson);
    case LESSON_ACTION_ERROR:
      return state.set("loading", false).set("error", action.error);
    case LESSON_COMPLETE_ACTION:
      return state.set("completeLoading", true).set("completeError", false);
    case LESSON_COMPLETE_ACTION_SUCCESS:
      return state
        .set("completeLoading", false)
        .set("currentLesson", action.lesson);
    case LESSON_COMPLETE_ACTION_ERROR:
      return state
        .set("completeLoading", false)
        .set("completeError", action.error);
    default:
      return state;
  }
}

export default appReducer;
