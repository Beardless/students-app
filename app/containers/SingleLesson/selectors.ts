import { createSelector } from "reselect";

const selectSingleLesson = state => state.get("singleLesson");

const makeSelectSingleLessonLoading = () =>
  createSelector(
    selectSingleLesson,
    lessonState => lessonState.get("loading")
  );

const makeSelectSingleLessonError = () =>
  createSelector(
    selectSingleLesson,
    lessonState => lessonState.get("error")
  );
const makeSelectSingleLesson = () =>
  createSelector(
    selectSingleLesson,
    lessonState => lessonState.get("currentLesson")
  );

export {
  selectSingleLesson,
  makeSelectSingleLessonLoading,
  makeSelectSingleLessonError,
  makeSelectSingleLesson
};
