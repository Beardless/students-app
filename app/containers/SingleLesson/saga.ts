import { LESSON_ACTION, LESSON_COMPLETE_ACTION } from "./constants";
import { call, put, takeEvery } from "redux-saga/effects";
import { getAuthorizationHeaders } from "../../utils/userUtils";
// @ts-ignore
import {
  lessonActionError,
  lessonActionSuccess,
  lessonCompleteActionError,
  lessonCompleteActionSuccess
} from "./actions";

// @ts-ignore
import request from "utils/request";
import { LessonInterface } from "./interfaces";

export function* getLesson({
  payload: { id: lessonId }
}: {
  payload: { id: string };
}) {
  try {
    const options = {
      method: "GET",
      url: `/lessons/${lessonId}`,
      headers: getAuthorizationHeaders()
    };
    const { data: lesson }: { data: LessonInterface } = yield call(
      request,
      options
    );

    yield put(lessonActionSuccess(lesson));
  } catch (err) {
    const { message } = err.response.data;
    yield put(lessonActionError(message));
  }
}

export function* completeLesson({
  payload: { isComplete, lessonId }
}: {
  payload: { isComplete: boolean; lessonId: string };
}) {
  try {
    const options = {
      method: "PATCH",
      url: `/lessons/${lessonId}/progress`,
      data: JSON.stringify({
        completed: isComplete
      }),
      headers: getAuthorizationHeaders()
    };

    const { data: lesson }: { data: LessonInterface } = yield call(
      request,
      options
    );

    yield put(lessonCompleteActionSuccess(lesson));
  } catch (err) {
    const { message } = err.response.data;
    yield put(lessonCompleteActionError(message));
  }
}

export default function* getSingleLesson() {
  // @ts-ignore
  yield takeEvery(LESSON_ACTION, getLesson);
  // @ts-ignore
  yield takeEvery(LESSON_COMPLETE_ACTION, completeLesson);
}
