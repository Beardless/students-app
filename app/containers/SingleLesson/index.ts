import reducer from "./reducer";
import saga from "./saga";
import { compose } from "redux";
import { connect } from "react-redux";
import SingleLesson from "./SingleLesson";
import injectSaga from "../../utils/injectSaga";
import injectReducer from "../../utils/injectReducer";
import { createStructuredSelector } from "reselect";
import { lessonAction } from "./actions";
import {
  makeSelectSingleLesson,
  makeSelectSingleLessonError,
  makeSelectSingleLessonLoading
} from "./selectors";

const mapStateToProps = createStructuredSelector({
  currentLesson: makeSelectSingleLesson(),
  error: makeSelectSingleLessonError(),
  loading: makeSelectSingleLessonLoading()
});

const mapDispatchToProps = dispatch => ({
  getLesson: id => dispatch(lessonAction(id))
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

const withReducer = injectReducer({ key: "singleLesson", reducer });
const withSaga = injectSaga({ key: "singleLesson", saga });

// @ts-ignore
export default compose(
  withSaga,
  withReducer,
  withConnect
)(SingleLesson);
