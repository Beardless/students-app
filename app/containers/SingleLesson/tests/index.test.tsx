import { shallow } from "enzyme";
import SingleLesson from "../SingleLesson";
import * as lesson from "./mocks/lesson.json";
import * as React from "react";
import { mapDispatchToProps } from "../index";
import { lessonAction } from "../actions";

describe("<SingleLesson />", () => {
  it("should render correctly", function() {
    const renderedComponent = shallow(
      <SingleLesson
        getLesson={() => {}}
        error={false}
        loading={false}
        currentLesson={lesson}
        match={{ params: { lessonId: "1" }, isExact: true, path: "", url: "" }}
      />
    );

    expect(renderedComponent).toMatchSnapshot();
  });

  describe("mapDispatchToProps", () => {
    describe("getLesson", () => {
      it("should be injected", function() {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.getLesson).toBeDefined();
      });

      it("should dispatch lessonAction when called", () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const id = "1";
        result.getLesson(id);
        expect(dispatch).toHaveBeenCalledWith(lessonAction(id));
      });
    });
  });
});
