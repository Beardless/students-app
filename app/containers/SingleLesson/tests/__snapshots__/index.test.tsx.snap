// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleLesson /> should render correctly 1`] = `
<div
  className="course-lesson-wrapper"
>
  <CourseLesson
    bookmarkProps={
      Object {
        "bookmarkError": false,
        "bookmarkList": null,
        "bookmarkLoading": false,
      }
    }
    completeLoading={false}
    deleteBookmark={[Function]}
    handleBookmarkClick={[Function]}
    item={
      Object {
        "attachments": Array [
          Object {
            "file_name": "attachment_example.pdf",
            "href": Object {
              "download": "example.com/example.png",
            },
            "id": "test_id",
            "mime_type": "application/pdf",
            "name": "dummy name",
          },
        ],
        "completed": null,
        "created": "2019-03-17T11:29:18+01:00",
        "description": "Test description of example lesson",
        "duration_in_minutes": 25,
        "embed_code": "Test embed_code of example lesson",
        "embed_type": "vimeo",
        "id": "53ca8a0a-b5a9-40e9-b5b7-0f9d226a4d8b",
        "module": Object {
          "course": Object {
            "id": "1",
            "title": "Test title of example course",
          },
          "id": "fb8fbc90-2a0f-45b3-83ee-bb6e9d1bc402",
          "title": "Test title of example module",
        },
        "position": 1,
        "title": "Test title of example lesson",
        "updated": "2019-03-17T12:24:44+01:00",
      }
    }
    onComplete={[Function]}
  />
  <BookmarkModal
    onCloseModal={[Function]}
    onSubmitModal={[Function]}
    open={false}
  />
</div>
`;
